### PR TITLE
[RELEASE] multisig entries for wallet api

### PR DESCRIPTION
**Pull request by naughtyfox**
Hey guys! We are making multisig gui wallet and we'd like these changes to be in upcoming release.

Added abilities in gui wallet api to:
* create multisig wallets
* multisig key images exchange
* create, sign, commit multisig transactions
* Refactored the facility to set Wallet status to work in atomic way.

Tested transaction sending on ubuntu 17.10
API usage example: https://github.com/naughtyfox/monero-multisig-api-example/blob/master/main.cpp

### CODE REVIEW

File: src/wallet/api/wallet.cpp [hunk 1]
```diff
@@ -1099,6 +1134,59 @@ bool WalletImpl::finalizeMultisig(const vector<string>& extraMultisigInfo) {
...
+        clearStatus();
+        checkMultisigWalletReady(m_wallet);
+
+        bool ready;
+        if (!m_wallet->multisig(&ready)) {
+            throw runtime_error("Wallet is not multisig");
+        }
+
+        if (!ready) {
+            throw runtime_error("Multisig wallet is not finalized yet");
```
Comment (stoffu) [inline src/wallet/api/wallet.cpp]: Seems spurious, since `checkMultisigWalletReady()` is called just above

File: src/wallet/api/pending_transaction.cpp [hunk 2]
```diff
@@ -78,7 +79,6 @@ std::vector<std::string> PendingTransactionImpl::txid() const
 
 bool PendingTransactionImpl::commit(const std::string &filename, bool overwrite)
 {
-
```
Comment (stoffu) [inline src/wallet/api/pending_transaction.cpp]: Spurious

File: src/wallet/wallet2.cpp [hunk 3]
```diff
@@ -4955,20 +4955,23 @@ bool wallet2::save_multisig_tx(const multisig_tx_set &txs, const std::string &fi
   return epee::file_io_utils::save_string_to_file(filename, ciphertext);
 }
 //----------------------------------------------------------------------------------------------------
-std::string wallet2::save_multisig_tx(const std::vector<pending_tx>& ptx_vector)
-{
-  multisig_tx_set txs;
-  txs.m_ptx = ptx_vector;
+wallet2::multisig_tx_set wallet2::make_multisig_tx_set(const std::vector<pending_tx>& ptx_vector) const {
+    multisig_tx_set txs;
```
Comment (stoffu) [inline src/wallet/wallet2.cpp]: Please conform to the same indentation convention already used in each file, which in this case is two whitespaces.

File: src/wallet/wallet2.cpp [hunk 4]
```diff
@@ -4955,20 +4955,23 @@ bool wallet2::save_multisig_tx(const multisig_tx_set &txs, const std::string &fi
...
-  for (const auto &msk: get_account().get_multisig_keys())
-  {
-    crypto::public_key pkey = get_multisig_signing_public_key(msk);
-    for (auto &ptx: txs.m_ptx) for (auto &sig: ptx.multisig_sigs) sig.signing_keys.insert(pkey);
-  }
+    for (const auto &msk: get_account().get_multisig_keys())
+    {
+      crypto::public_key pkey = get_multisig_signing_public_key(msk);
+      for (auto &ptx: txs.m_ptx) for (auto &sig: ptx.multisig_sigs) sig.signing_keys.insert(pkey);
+    }
```
Comment (stoffu) [inline src/wallet/wallet2.cpp]: No whitespace changes

File: src/wallet/wallet2.h [hunk 5]
```diff
@@ -821,6 +821,8 @@ namespace tools
       a & m_ring_history_saved;
     }
 
+    multisig_tx_set make_multisig_tx_set(const std::vector<pending_tx>& ptx_vector) const;
```
Comment (stoffu) [inline src/wallet/wallet2.h]: Maybe move this line next to `save_multisig_tx()` as they're closely related?

File: src/wallet/api/wallet.cpp [hunk 6]
```diff
@@ -1195,6 +1195,34 @@ size_t WalletImpl::importMultisigImages(const vector<string>& images) {
...
+        clearStatus();
+        checkMultisigWalletReady(m_wallet);
+
+        string binary;
+        if (!epee::string_tools::parse_hexstr_to_binbuff(signData, binary)) {
+            throw runtime_error("Failed to deserialize multisig transaction");
+        }
+
+        tools::wallet2::multisig_tx_set txSet;
+        if (!m_wallet->load_multisig_tx(binary, txSet, std::function<bool(const tools::wallet2::multisig_tx_set&)>())) {
```
Comment (stoffu) [inline src/wallet/api/wallet.cpp]: The last argument can simply be `{}`

File: src/wallet/api/wallet2_api.h [hunk 7]
```diff
@@ -628,6 +663,48 @@ struct Wallet
...
+    virtual MultisigState multisig() const = 0;
+    /**
+     * @brief getMultisigInfo
+     * @return serialized and signed multisig info string
+     */
+    virtual std::string getMultisigInfo() const = 0;
+    /**
+     * @brief makeMultisig - switches wallet in multisig state. The one and only creation phase for N / N wallets
+     * @param info - vector of multisig infos from other participants obtained with getMulitisInfo call
+     * @param threshold - number of required signatures to make valid transaction. Must be equal to number of participants (N) or N - 1
```
Comment (stoffu) [inline src/wallet/api/wallet2_api.h]: I'd rephrase "number of required signatures" to "number of required signers".

File: src/wallet/api/wallet.cpp [hunk 8]
```diff
@@ -1306,89 +1253,70 @@ PendingTransaction *WalletImpl::createSweepUnmixableTransaction()
...
-            m_status = Status_Error;
+            setStatusError(tr("RPC error: ") +  e.to_string());
         } catch (const tools::error::get_random_outs_error&) {
-            m_errorString = tr("failed to get random outputs to mix");
-            m_status = Status_Error;
-
+            setStatusError(tr("failed to get random outputs to mix"));
         } catch (const tools::error::not_enough_unlocked_money& e) {
-            m_status = Status_Error;
+            setStatusError("");
```
[the remark this answers is not in the document]
Comment (naughtyfox) [reply src/wallet/api/wallet.cpp]: actually no. in previous version of code there was something like this:
```
m_status = Status_Error;
....
m_status = Status_Error;
m_errorString = "...";
```
because api lib wanted a caller to know error happened before setting actual error string. I followed the same way since `m_status` was atomic

File: src/wallet/api/wallet.cpp [hunk 9]
```diff
@@ -1306,89 +1253,70 @@ PendingTransaction *WalletImpl::createSweepUnmixableTransaction()
...
 
             writer << boost::format(tr("not enough money to transfer, available only %s, sent amount %s")) %
                       print_money(e.available()) %
                       print_money(e.tx_amount());
-            m_errorString = writer.str();
-
+            setStatusError(writer.str());
         } catch (const tools::error::not_enough_money& e) {
-            m_status = Status_Error;
+            setStatusError("");
```
[the remark this answers is not in the document]
Comment (naughtyfox) [reply src/wallet/api/wallet.cpp]: answered above

File: src/wallet/api/wallet.cpp (hunk 1 above)
Comment (naughtyfox) [reply src/wallet/api/wallet.cpp]: you absolutely right, thank you! i wrote this code first then refactored a bit and forgot about this lines. will fix it

File: src/wallet/api/pending_transaction.cpp (hunk 2 above)
Comment (naughtyfox) [reply src/wallet/api/pending_transaction.cpp]: thought it unnecessary :) bring it back?

File: src/wallet/api/wallet2_api.h (hunk 7 above)
Comment (naughtyfox) [reply src/wallet/api/wallet2_api.h]: good point, i'll do it

File: src/wallet/api/wallet.cpp (hunk 6 above)
Comment (naughtyfox) [reply src/wallet/api/wallet.cpp]: that's it! thx :)

File: src/wallet/api/pending_transaction.cpp (hunk 2 above)
Comment (stoffu) [reply src/wallet/api/pending_transaction.cpp]: This kind of cosmetic changes are generally discouraged. Minimizing the number of changed lines is much more important.

File: src/wallet/api/wallet.cpp (hunk 1 above)
Comment (naughtyfox) [reply src/wallet/api/wallet.cpp]: i deleted redundant checks but github shows it still for some reason

File: src/wallet/api/pending_transaction.cpp (hunk 2 above)
Comment (naughtyfox) [reply src/wallet/api/pending_transaction.cpp]: ok, thank you. i'll take this into account in future

**Comment by stoffu** [comment]
Please do not make changes to indentations if there's no functional change.

**Comment by naughtyfox** [comment]
Rebased commits but github it doubles for some mysterious reason
@stoffu check this out please

**Comment by naughtyfox** [comment]
rebased commits 
@stoffu please check this out

**Comment by stoffu** [comment]
Are you intentionally refusing to remove those indentation changes and spurious line addition/deletion, which I'm _strongly_ against?

**Comment by naughtyfox** [comment]
@stoffu Sorry, just got your message about minimizing changes wrong. Brought it back

File: src/wallet/wallet2.cpp [hunk 10]
```diff
@@ -4955,18 +4955,19 @@ bool wallet2::save_multisig_tx(const multisig_tx_set &txs, const std::string &fi
   return epee::file_io_utils::save_string_to_file(filename, ciphertext);
 }
 //----------------------------------------------------------------------------------------------------
-wallet2::multisig_tx_set wallet2::make_multisig_tx_set(const std::vector<pending_tx>& ptx_vector) const {
```
Comment (stoffu) [inline src/wallet/wallet2.cpp]: Please revert this.

File: src/wallet/wallet2.h [hunk 11]
```diff
@@ -821,8 +822,6 @@ namespace tools
       a & m_ring_history_saved;
     }
 
-    multisig_tx_set make_multisig_tx_set(const std::vector<pending_tx>& ptx_vector) const;
-
```
Comment (stoffu) [inline src/wallet/wallet2.h]: Please squash this change to the previous commit where this function was introduced.

File: src/wallet/wallet2.cpp (hunk 10 above)
Comment (naughtyfox) [reply src/wallet/wallet2.cpp]: why do you want to revert it? this function is used in `WalletImpl::createTransaction`

**Comment by stoffu** [comment]
There should be just two commits. The last two commits should be squashed into the second commit, and the indentation change should be removed.

**Comment by naughtyfox** [comment]
i thought i removed all of identations... did i miss something?

**Comment by stoffu** [comment]
Don't you see the problem in the color-coded diff that GitHub offers?

**Comment by naughtyfox** [comment]
Done

**Comment by naughtyfox** [comment]
Squashed into 2 commits:
* refactoring
* feature commit

**Comment by naughtyfox** [comment]
No worries! It's my fault i didn't read the contribution doc, that's my first open source contribution and i've been a bit distracted. Thank you for your patience!

File: src/wallet/api/pending_transaction.cpp [hunk 12]
```diff
@@ -188,6 +194,53 @@ std::vector<std::set<uint32_t>> PendingTransactionImpl::subaddrIndices() const
...
+std::string PendingTransactionImpl::multisigSignData() {
+    try {
+        if (!m_wallet.multisig().isMultisig) {
+            throw std::runtime_error("wallet is not multisig");
+        }
+
+        auto cipher = m_wallet.m_wallet->save_multisig_tx(m_pending_tx);
+        return epee::string_tools::buff_to_hex_nodelimer(cipher);
+    } catch (const std::exception& e) {
+        m_status = Status_Error;
```
Comment (moneromooo-monero) [inline src/wallet/api/pending_transaction.cpp]: Should this not lock the new mutex (and others below) ?

Comment (naughtyfox) [reply src/wallet/api/pending_transaction.cpp]: It seems `PendingTransactionImpl` class is not thread safe and there's no mutex at all. I just followed the same logic. If you want i can fix it

**Comment by moneromooo-monero** [comment]
No it's fine, I did not know it was a different m_status field.

**Comment by youmind** [comment]
jesli możesz napraw
Dawid Rafał Michalski d.michalski654@gmail.com


śr., 30 maj 2018 o 13:21 naughtyfox <notifications@github.com> napisał(a):
